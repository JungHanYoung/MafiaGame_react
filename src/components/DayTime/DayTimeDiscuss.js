import React from 'react';
import PropTypes from 'prop-types';

const DayTimeDiscuss = ({ changeDayTimeOrder }) => (
	<main className="daytime">
		<h2 className="game-title">hello mafia</h2>
		<div className="game-content">
			<p className="content-description">낮 토론 시간입니다.
				<br /> 토론을 통해<br /> 마피아를 찾으세요.</p>
		</div>
		<button
			className="btn-lg"
			onClick={changeDayTimeOrder}>투표로 이동</button>

	</main>
)

DayTimeDiscuss.propTypes = {
	changeDayTimeOrder: PropTypes.func.isRequired
};

export default DayTimeDiscuss
