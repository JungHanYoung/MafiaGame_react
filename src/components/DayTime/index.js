import React from 'react';

// Component
import Discuss from './DayTimeDiscuss';
import VoteTime from './VoteTime';
import Result from './Result';
import { useGame } from '../../context/GameContext';

class DayTime extends React.Component {
	handleNext = () => {
		const { dayTimeOrder, changeDayTimeOrder, toggleNightAndDay } = this.props;
		if (dayTimeOrder === 'result') {
			toggleNightAndDay();
		}
		changeDayTimeOrder();
	};
	render() {
		const { dayTimeOrder, isEndGame, moveToMain } = this.props;
		return (
			<div>
				{dayTimeOrder === 'discuss' && <Discuss />}
				{dayTimeOrder === 'vote' && <VoteTime />}
				{dayTimeOrder === 'result' && <Result />}
				{dayTimeOrder !== 'vote' && !isEndGame && <button onClick={this.handleNext}>next</button>}
				{isEndGame && <button onClick={moveToMain}>메인으로</button>}
			</div>
		);
	}
}

export default useGame(({ state, actions }) => ({
	isEndGame: state.isEndGame,
	dayTimeOrder: state.dayTimeOrder,
	moveToMain: actions.moveToMain,
	changeDayTimeOrder: actions.changeDayTimeOrder,
	toggleNightAndDay: actions.toggleNightAndDay
}))(DayTime);
