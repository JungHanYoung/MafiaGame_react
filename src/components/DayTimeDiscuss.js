import React from 'react';

class DayTimeDiscuss extends React.Component {
	render() {
		return <div>낮 토론 시간입니다. 토론을 통해 마피아를 찾으세욥.</div>;
	}
}

export default DayTimeDiscuss;
