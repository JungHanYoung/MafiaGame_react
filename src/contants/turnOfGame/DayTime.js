export const TURN_OF_DISCUSS_AT_DAY = 'discuss';
export const TURN_OF_VOTE_AT_DAY = 'vote';
export const TURN_OF_RESULT_AT_DAY = 'result';
