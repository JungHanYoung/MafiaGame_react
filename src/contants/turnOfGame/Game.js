export const DAY_TIME = 'day-time';
export const NIGHT_TIME = 'night-time';
