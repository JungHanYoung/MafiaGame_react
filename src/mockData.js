import { List, Map } from "immutable";
import { JOB_NAME_OF_CITIZEN, JOB_NAME_OF_POLICE, JOB_NAME_OF_DOCTOR, JOB_NAME_OF_MAFIA } from "./contants/Job";


export const players = List([
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 3
    }),
    Map({
        name: "병욱",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "지은",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 0
    }),
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    })
])

const playerWithoutImmutable = [
    {
        name: '한영',
        jobName: JOB_NAME_OF_CITIZEN
    },
    {
        name: '병욱',
        jobName: JOB_NAME_OF_MAFIA
    },
    {
        name: '대용',
        jobName: JOB_NAME_OF_DOCTOR
    },
    {
        name: '기수',
        jobName: JOB_NAME_OF_POLICE
    },
]

export const minPropsByGame = {
    history: {
        location: {
            state: {
                players: playerWithoutImmutable
            }
        }
    }
}

export const playersWithMafia = List([
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 0
    }),
    Map({
        name: "병욱",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "지은",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 0
    }),
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "민기",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "하림",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
])