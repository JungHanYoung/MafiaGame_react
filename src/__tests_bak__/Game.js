import React from 'react';
import { shallow, mount } from "enzyme";
import { List, Map } from 'immutable'

// Component
import Game from '../routes/Game'

// props
import { minPropsByGame, players } from '../mockData'
import { DAY_TIME } from '../contants/turnOfGame/Game';
import DayTime from '../components/DayTime';
import DayTimeDiscuss from '../components/DayTime/DayTimeDiscuss';


describe('<Game /> 컴포넌트', () => {

    const wrapper = shallow(<Game {...minPropsByGame} />)

    it('렌더링 테스트', () => {
        expect(wrapper).toBeDefined()
    })

    it('state 세팅 테스트', () => {
        const state = wrapper.state()
        expect(state).toHaveProperty('gameOrder');
        expect(state).toHaveProperty('players');
        expect(state).toHaveProperty('playersByResult');

        expect(state.gameOrder).toBe(DAY_TIME)
        expect(state.players).toBeInstanceOf(List)
        state.players.forEach(player => {
            expect(player).toBeInstanceOf(Map)
            const playerToJS = player.toJS()
            expect(playerToJS).toHaveProperty('name')
            expect(playerToJS).toHaveProperty('jobName')
            expect(playerToJS).toHaveProperty('daytimeVoted')
        })
        expect(state.playersByResult).toBeInstanceOf(Array)
        state.playersByResult.forEach(player => {
            expect(player).toBeInstanceOf(Object)
            expect(player).toHaveProperty('name')
            expect(player).toHaveProperty('jobName')
        })
    })

    it('html 렌더링 확인', () => {
        expect(wrapper.find('main.daytime')).toBeDefined()
        expect(wrapper.find('h2.game-title')).toBeDefined()
        // expect(wrapper.find('h2.game-title').text()).toEqual('HELLO MAFIA')
    })

    it('DayTime 렌더링', () => {

        expect(wrapper.find(DayTime).exists()).toBe(true)
    })

    it('DayTime > Discuss 렌더링', () => {

        expect(wrapper.find(DayTime).shallow().find(DayTimeDiscuss).exists()).toBe(true)
    })

})

describe('Game 컴포넌트 >> mount', () => {

    const wrapper = mount(<Game {...minPropsByGame} />)

    it('클릭', () => {
        wrapper.find('button.btn-lg').simulate('click')
        expect(wrapper.find('p.content-description').text()).toBe('마피아로 의심되는 사람을 투표합니다.')
        const voteButtons = wrapper.find('button.btn-sm')
        expect(voteButtons).toHaveLength(3)
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')

        expect(wrapper.find('p.content-description').text()).toContain('님이 죽으셨습니다.')
        expect(wrapper.find('button.btn-lg').text()).toBe('밤이 됩니다.')

        wrapper.find('button.btn-lg').simulate('click')

        expect(wrapper.find('main.night').exists()).toBe(true)
    })
})

test('컴포넌트 메소드', () => {
    const wrapper = shallow(<Game {...minPropsByGame} />)
    wrapper.instance().deletePlayer(null)

    expect(wrapper.state()).toMatchSnapshot()
})

// describe('컴포넌트 메소드', () => {
//     const minPropsWithResult = {

//     }
// })
test('players findIndex Test', () => {
    const index = players.findIndex(person => person.get('name') === null)
    expect(index).toBe(-1);

})

test('set test', () => {

})