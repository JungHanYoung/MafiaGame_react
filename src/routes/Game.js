import React, { useState, useEffect } from 'react';
import { List, Map } from 'immutable'
import { Redirect } from 'react-router-dom'
// import PropTypes from 'prop-types';

// 컴포넌트
import DayTime from '../components/DayTime';
import NightTime from '../components/NightTime';

// constant
import { DAY_TIME, NIGHT_TIME } from '../contants/turnOfGame/Game';

export default function Game(props) {
	const [players, setPlayers] = useState(
		List(
			props.history.location.state.players
				.map(player => Map(player).set('daytimeVoted', 0))));
	const [gameOrder, setGameOrder] = useState(DAY_TIME);

	useEffect(() => {
		initVote()
	}, [gameOrder])

	function moveToResult() {
		props.history.push('/result', { players: props.history.location.state.players })
	}

	function changeDayAndNight() {
		setGameOrder(gameOrder === DAY_TIME ? NIGHT_TIME : DAY_TIME);
	}

	function votePerson(name) {
		setPlayers(
			players.updateIn([players.findIndex(player => player.get('name') === name), 'daytimeVoted'], val => val + 1))
	}

	function deletePlayer(name) {
		setPlayers(
			players.delete(players.findIndex(player => player.get('name') === name))
		)
	}

	function initVote() {
		setPlayers(
			players.map(player => player.set('daytimeVoted', 0))
		)
	}

	if (players.size) {
		return <>
			{gameOrder === DAY_TIME ? (
				<DayTime
					players={players}
					initVote={initVote}
					votePerson={votePerson}
					changeDayAndNight={changeDayAndNight}
					deletePlayer={deletePlayer}
					moveToResult={moveToResult}
				/>
			) : (
					<main className="night animated fadeInDown">
						<h2 className="game-title">HELLO MAFIA</h2>
						<NightTime
							players={players}
							changeDayAndNight={changeDayAndNight}
							deletePlayer={deletePlayer}
							moveToResult={moveToResult}
						/>
					</main>
				)}
		</>
	} else {
		return <Redirect to="/setting" />
	}
}

// class Game extends React.Component {
// 	constructor(props) {
// 		super(props)

// 		const { history } = props;
// 		const players = history.location.state.players
// 		this.state = {
// 			players: List(players.map(player => Map(player).set('daytimeVoted', 0))),
// 			playersByResult: players,
// 			gameOrder: DAY_TIME
// 		}
// 	}

// 	moveToMain = () => {
// 		this.props.history.push('/')
// 	}

// 	moveToResult = () => {
// 		this.props.history.push('/result', { players: this.state.playersByResult })
// 	}

// 	initVote = () => {
// 		const { players } = this.state;
// 		this.setState({
// 			players: players.map(player => player.set('daytimeVoted', 0))
// 		})
// 	}

// 	render() {
// 		const { players, gameOrder } = this.state;
// 		if (players.size) {
// 			return <>
// 				{gameOrder === DAY_TIME ? (
// 					<DayTime
// 						players={players}
// 						initVote={this.initVote}
// 						votePerson={this.votePerson}
// 						changeDayAndNight={this.changeDayAndNight}
// 						deletePlayer={this.deletePlayer}
// 						moveToResult={this.moveToResult}
// 					/>
// 				) : (
// 						<main className="night animated fadeInDown">
// 							<h2 className="game-title">HELLO MAFIA</h2>
// 							<NightTime
// 								players={players}
// 								changeDayAndNight={this.changeDayAndNight}
// 								deletePlayer={this.deletePlayer}
// 								moveToResult={this.moveToResult}
// 							/>
// 						</main>
// 					)}
// 			</>
// 		} else {
// 			return <Redirect to="/setting" />
// 		}
// 	}

// 	changeDayAndNight = () => {
// 		this.setState(state => ({
// 			gameOrder: state.gameOrder === DAY_TIME ?
// 				NIGHT_TIME : DAY_TIME
// 		}))
// 	}

// 	votePerson = (name) => {
// 		const { players } = this.state;
// 		this.setState({
// 			players: players.updateIn([players.findIndex(player => player.get('name') === name), 'daytimeVoted'], val => val + 1)
// 		})
// 	}

// 	deletePlayer = (name) => {
// 		const { players } = this.state;
// 		console.log(name);
// 		this.setState({
// 			players: players.delete(players.findIndex(player => player.get('name') === name))
// 		})
// 	}
// }

// export default Game;
