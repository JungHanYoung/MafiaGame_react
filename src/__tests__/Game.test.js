import React from 'react';
import { shallow, mount } from "enzyme";
import { List, Map } from 'immutable'

// Component
import Game from '../routes/Game'

// props
import { minPropsByGame } from '../mockData'
import { DAY_TIME } from '../contants/turnOfGame/Game';
import DayTime from '../components/DayTime';
import DayTimeDiscuss from '../components/DayTime/DayTimeDiscuss';


describe('<Game /> 컴포넌트', () => {

    const wrapper = shallow(<Game {...minPropsByGame} />)

    it('렌더링 테스트', () => {
        expect(wrapper).toBeDefined()
    })

    it('state 세팅 테스트', () => {
        expect(wrapper.state()).toMatchSnapshot()
    })

    it('html 렌더링 확인', () => {
        expect(wrapper.children()).toMatchSnapshot()
        expect(wrapper.find('main.daytime')).toBeDefined()
        expect(wrapper.find('h2.game-title')).toBeDefined()
        expect(wrapper.find('h2.game-title').text()).toEqual('HELLO MAFIA')
    })

    it('DayTime 렌더링', () => {

        expect(wrapper.find(DayTime).shallow()).toMatchSnapshot()
    })

    it('DayTime > Discuss 렌더링', () => {

        expect(wrapper.find(DayTime).shallow().find(DayTimeDiscuss).shallow()).toMatchSnapshot()
        expect(wrapper.props()).toMatchSnapshot()
    })

})

describe('Game 컴포넌트 >> mount', () => {

    const wrapper = mount(<Game {...minPropsByGame} />)

    it('스냅샷', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('클릭', () => {
        wrapper.find('button.btn-lg').simulate('click')

        expect(wrapper).toMatchSnapshot()

        expect(wrapper.find('div.vote-btn-container').children().find('button.btn-sm')).toMatchSnapshot()

        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')
        wrapper.find('div.vote-btn-container').children().find('button.btn-sm').first().simulate('click')

        expect(wrapper).toMatchSnapshot()

        wrapper.find('button.btn-lg').simulate('click')

        expect(wrapper).toMatchSnapshot()
    })
})