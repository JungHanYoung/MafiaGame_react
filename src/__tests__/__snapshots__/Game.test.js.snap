// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> 컴포넌트 DayTime > Discuss 렌더링 1`] = `
<Fragment>
  <div
    className="game-content"
  >
    <h2>
      낮 토론 시간입니다.
      <br />
       토론을 통해 마피아를 찾으세요.
    </h2>
  </div>
  <button
    className="btn-lg"
    onClick={[Function]}
  >
    투표로 이동
  </button>
</Fragment>
`;

exports[`<Game /> 컴포넌트 DayTime > Discuss 렌더링 2`] = `
Object {
  "children": <main
    className="daytime"
  >
    <h2
      className="game-title"
    >
      HELLO MAFIA
    </h2>
    <DayTime
      changeDayAndNight={[Function]}
      deletePlayer={[Function]}
      moveToMain={[Function]}
      players={
        Immutable.List [
          Immutable.Map {
            "name": "한영",
            "jobName": "CITIZEN",
            "daytimeVoted": 0,
          },
          Immutable.Map {
            "name": "병욱",
            "jobName": "MAFIA",
            "daytimeVoted": 0,
          },
          Immutable.Map {
            "name": "대용",
            "jobName": "DOCTOR",
            "daytimeVoted": 0,
          },
          Immutable.Map {
            "name": "기수",
            "jobName": "POLICE",
            "daytimeVoted": 0,
          },
        ]
      }
      votePerson={[Function]}
    />
  </main>,
}
`;

exports[`<Game /> 컴포넌트 DayTime 렌더링 1`] = `
<Fragment>
  <DayTimeDiscuss
    changeDayTimeOrder={[Function]}
  />
</Fragment>
`;

exports[`<Game /> 컴포넌트 html 렌더링 확인 1`] = `
<main
  className="daytime"
>
  <h2
    className="game-title"
  >
    HELLO MAFIA
  </h2>
  <DayTime
    changeDayAndNight={[Function]}
    deletePlayer={[Function]}
    moveToMain={[Function]}
    players={
      Immutable.List [
        Immutable.Map {
          "name": "한영",
          "jobName": "CITIZEN",
          "daytimeVoted": 0,
        },
        Immutable.Map {
          "name": "병욱",
          "jobName": "MAFIA",
          "daytimeVoted": 0,
        },
        Immutable.Map {
          "name": "대용",
          "jobName": "DOCTOR",
          "daytimeVoted": 0,
        },
        Immutable.Map {
          "name": "기수",
          "jobName": "POLICE",
          "daytimeVoted": 0,
        },
      ]
    }
    votePerson={[Function]}
  />
</main>
`;

exports[`<Game /> 컴포넌트 state 세팅 테스트 1`] = `
Object {
  "gameOrder": "day-time",
  "players": Immutable.List [
    Immutable.Map {
      "name": "한영",
      "jobName": "CITIZEN",
      "daytimeVoted": 0,
    },
    Immutable.Map {
      "name": "병욱",
      "jobName": "MAFIA",
      "daytimeVoted": 0,
    },
    Immutable.Map {
      "name": "대용",
      "jobName": "DOCTOR",
      "daytimeVoted": 0,
    },
    Immutable.Map {
      "name": "기수",
      "jobName": "POLICE",
      "daytimeVoted": 0,
    },
  ],
}
`;

exports[`Game 컴포넌트 >> mount 스냅샷 1`] = `
<main
  className="daytime"
>
  <h2
    className="game-title"
  >
    HELLO MAFIA
  </h2>
  Array [
    <div
      className="game-content"
    >
      <h2>
        낮 토론 시간입니다.
        <br />
         토론을 통해 마피아를 찾으세요.
      </h2>
    </div>,
    <button
      className="btn-lg"
      onClick={[Function]}
    >
      투표로 이동
    </button>,
  ]
</main>
`;

exports[`Game 컴포넌트 >> mount 클릭 1`] = `
<main
  className="daytime"
>
  <h2
    className="game-title"
  >
    HELLO MAFIA
  </h2>
  <div
    className="game-content"
  >
    <p
      className="content-description"
    >
      마피아로 의심되는 사람을 투표합니다.
    </p>
    <div
      className="voter"
    >
      한영
      님의 투표
    </div>
    <div
      className="vote-btn-container"
    >
      <div>
        <button
          className="btn-sm"
          key="vote-btn-병욱"
          onClick={[Function]}
        >
          병욱
        </button>
        <button
          className="btn-sm"
          key="vote-btn-대용"
          onClick={[Function]}
        >
          대용
        </button>
        <button
          className="btn-sm"
          key="vote-btn-기수"
          onClick={[Function]}
        >
          기수
        </button>
      </div>
    </div>
  </div>
</main>
`;

exports[`Game 컴포넌트 >> mount 클릭 2`] = `
Array [
  <button
    className="btn-sm"
    key="vote-btn-병욱"
    onClick={[Function]}
  >
    병욱
  </button>,
  <button
    className="btn-sm"
    key="vote-btn-대용"
    onClick={[Function]}
  >
    대용
  </button>,
  <button
    className="btn-sm"
    key="vote-btn-기수"
    onClick={[Function]}
  >
    기수
  </button>,
]
`;

exports[`Game 컴포넌트 >> mount 클릭 3`] = `
<main
  className="daytime"
>
  <h2
    className="game-title"
  >
    HELLO MAFIA
  </h2>
  <div
    className="game-content"
  >
    <h2>
      투표 결과
    </h2>
    <p
      className="content-description"
    >
      <span
        style={
          Object {
            "color": "#ff0000",
          }
        }
      >
        한영
      </span>
       님이 죽으셨습니다.
    </p>
    <button
      className="btn-lg"
      onClick={[Function]}
    >
      밤이 됩니다.
    </button>
  </div>
</main>
`;

exports[`Game 컴포넌트 >> mount 클릭 4`] = `
<main
  className="night"
>
  <h1>
    HELLO MAFIA
  </h1>
  Array [
    <div
      className="game-content"
    >
      <h1>
        병욱
        의 차례입니다.
      </h1>
    </div>,
    <button
      className="btn-lg"
      onClick={[Function]}
    >
      역할 확인
    </button>,
  ]
</main>
`;
